utils/s3/client.cc

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <rapidxml.h>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/util/short_streams.hh>
+#include <seastar/http/request.hh>
+#include "utils/s3/client.hh"
+#include "utils/memory_data_sink.hh"
+#include "utils/chunked_vector.hh"
+#include "log.hh"
+
+namespace s3 {
+
+static logging::logger s3l("s3");
+
+future<> ignore_reply(http::reply rep, input_stream<char>&& in_) {
+    auto in = std::move(in_);
+    co_await util::skip_entire_stream(in);
+}
+
+client::client(ipv4_addr addr, private_tag)
+        : _host(to_sstring(addr))
+        , _http(std::move(addr))
+{
+}
+
+shared_ptr<client> client::make(ipv4_addr addr) {
+    return seastar::make_shared<client>(std::move(addr), private_tag{});
+}
+
+future<uint64_t> client::get_object_size(sstring object_name) {
+    s3l.trace("HEAD {}", object_name);
+    auto req = http::request::make("HEAD", _host, object_name);
+    uint64_t len = 0;
+    co_await _http.make_request(std::move(req), [&len] (http::reply rep, input_stream<char>&& in_) mutable -> future<> {
+        len = rep.content_length;
+        return make_ready_future<>(); // it's HEAD with no body
+    });
+    co_return len;
+}
+
+future<temporary_buffer<char>> client::get_object_contiguous(sstring object_name, std::optional<range> range) {
+    s3l.trace("GET {} contiguous", object_name);
+    auto req = http::request::make("GET", _host, object_name);
+    http::reply::status_type expected = http::reply::status_type::ok;
+    if (range) {
+        req._headers["Range"] = format("bytes={}-{}", range->off, range->off + range->len - 1);
+        expected = http::reply::status_type::partial_content;
+    }
+
+    size_t off = 0;
+    std::optional<temporary_buffer<char>> ret;
+    co_await _http.make_request(std::move(req), [&off, &ret, &object_name] (http::reply rep, input_stream<char>&& in_) mutable -> future<> {
+        auto in = std::move(in_);
+        ret = temporary_buffer<char>(rep.content_length);
```

Comment:
We will want to add (later) a customization point so we can wrap this with a make_tracked_temporary_buffer(). We may also want to adjust the sstable reader to be able to consume buffers as they are read in, rather than wait+copy. The latter change is very large for dubious gains, so only after we gain some experience with it.

Reply:
Please create an issue and link it into the main object storage tracker so we don't forget.

Reply:
#13306